File: core/contractsdk/cpp/example/luck_draw.cc
#include "xchain/xchain.h"

const std::string TICKETID = "Luckid_";
const std::string USERID = "Userid_";
const std::string ADMIN = "admin";
const std::string RESULT = "result";
const std::string TICKETS = "tickets";

// 抽奖小游戏模板
// 参数由xchain::Contract中的context提供
class LuckDraw {
public:
    /*
     * func: 初始化游戏
     * @param: admin: 哪个address具有管理员权限
     */
    virtual void initialize() = 0;
    /*
     * func: 获得一个抽奖券
     * @param: initiator: 玩家的address，获得一个抽奖券
     */
    virtual void getLuckid() = 0;
    /*
     * func: 开始抽奖
     * @param: seed:
     * 传入一个随机数种子，可以是预言机生成的，也可以是游戏约定的，例如某天A股收盘价
     */
    virtual void startLuckDraw() = 0;
    /*
     * func: 查询抽奖结果
     */
    virtual void getResult() = 0;
};

struct LuckDrawDemo : public LuckDraw, public xchain::Contract {
public:
    void initialize() {
        xchain::Context* ctx = this->context();
        const std::string& admin = ctx->arg(ADMIN);
        if (admin.empty()) {
            ctx->error("missing admin address");
            return;
        }

        std::string key = ADMIN;
        ctx->put_object(key, admin);
        ctx->put_object(TICKETS, "0");
        ctx->ok("initialize success");
    }

    bool isAdmin(xchain::Context* ctx, const std::string& caller) {
        std::string admin;
        if (!ctx->get_object(ADMIN, &admin)) {
            return false;
        }
        return (admin == caller);
    }

    void getLuckid() {
        xchain::Context* ctx = this->context();
        const std::string& caller = ctx->initiator();
        if (caller.empty()) {
            ctx->error("missing initiator");
            return;
        }

        // 检查是否存在抽奖结果，如果存在则不再继续发放奖券
        std::string result;
        if (ctx->get_object(RESULT, &result)) {
            ctx->error("this luck draw is finished");
            return;
        }

        // 检查用户是否已经抽过奖券, 如果抽过直接返回上次抽的奖券号
        std::string userval;
        if (ctx->get_object(USERID + caller, &userval)) {
            ctx->ok(userval);
            return;
        }

        std::string lastidStr;
        if (!ctx->get_object(TICKETS, &lastidStr)) {
            ctx->error("get tickets count failed");
            return;
        }

        // 确定当前抽奖券编号，全局自增
        int lastid = std::atoi(lastidStr.c_str());
        if (lastid < 0) {
            ctx->error("tickets count is wrong");
            return;
        }
        lastid++;
        lastidStr = std::to_string(lastid);
        if (!ctx->put_object(USERID + caller, lastidStr) ||
            !ctx->put_object(TICKETID + lastidStr, caller) ||
            !ctx->put_object(TICKETS, lastidStr)) {
            ctx->error("save ticket failed");
            return;
        }
        ctx->ok(lastidStr);
    }

    void startLuckDraw() {
        xchain::Context* ctx = this->context();
        const std::string& caller = ctx->initiator();
        if (caller.empty()) {
            ctx->error("missing initiator");
            return;
        }

        if (!isAdmin(ctx, caller)) {
            ctx->error("only the admin can add new asset type");
            return;
        }

        const std::string& seedStr = ctx->arg("seed");
        if (seedStr.empty()) {
            ctx->error("missing seed");
            return;
        }
        int seed = std::atoi(seedStr.c_str());

        // 获取总奖券数
        std::string lastidStr;
        if (!ctx->get_object(TICKETS, &lastidStr)) {
            ctx->error("get tickets count failed");
            return;
        }
        int lastid = std::atoi(lastidStr.c_str());
        if (lastid == 0) {
            ctx->error("no luck draw tickets");
            return;
        }

        // 抽奖
        srand(seed);
        int lucknum = (rand() % lastid) + 1;
        std::string luckid = std::to_string(lucknum);

        std::string luckuser;
        if (!ctx->get_object(TICKETID + luckid, &luckuser)) {
            ctx->error("get luck ticket failed");
            return;
        }

        // 记录抽奖结果
        if (!ctx->put_object(RESULT, luckuser)) {
            ctx->error("save luck draw result failed");
            return;
        }
        ctx->ok(luckuser);
    }

    void getResult() {
        xchain::Context* ctx = this->context();
        // 获取总奖券数
        std::string luckuser;
        if (!ctx->get_object(RESULT, &luckuser)) {
            ctx->error("get luck draw result failed");
            return;
        }

        ctx->ok(luckuser);
    };
};

DEFINE_METHOD(LuckDrawDemo, initialize) { self.initialize(); }

DEFINE_METHOD(LuckDrawDemo, getLuckid) { self.getLuckid(); }

DEFINE_METHOD(LuckDrawDemo, startLuckDraw) { self.startLuckDraw(); }

DEFINE_METHOD(LuckDrawDemo, getResult) { self.getResult(); }

File: core/contractsdk/java/example/luck_draw/src/main/java/com/baidu/xuper/LuckDraw.java
package com.baidu.xuper;

import com.sun.tools.javac.util.ByteBuffer;

import java.math.BigDecimal;
import java.util.Arrays;
import java.util.Random;

/**
 * Counter
 */
public class LuckDraw implements Contract {
    final String admin_key = "admin";
    final String userID = "result";
    final String result = "tickets";
    final String ticketId = "userid";
    final String tickets = "tickets";

    @Override
    @ContractMethod
    public Response initialize(Context ctx) {
        String admin = new String(ctx.args().get(admin_key));
        if (admin.length() == 0) {
            return Response.error("missing admin address");
        }
        ctx.putObject(admin.getBytes(), admin_key.getBytes());

        return Response.ok("ok".getBytes());
    }

    @ContractMethod
    public Response getLuckId(Context ctx) {
        String caller = ctx.caller();
        if (caller.isEmpty()) {
            return Response.error("missing initiator");
        }
        if (ctx.getObject(result.getBytes()).length != 0) {
            return Response.error("this luck draw has finished");
        }
        byte[] userVal = ctx.getObject((userID + caller).getBytes());
        if (userVal.length > 0) {
            return Response.ok(null);
        }
        BigDecimal lastId = new BigDecimal(new String(ctx.getObject(tickets.getBytes())));
        lastId = lastId.add(BigDecimal.ONE);
        ctx.putObject((userID + caller).getBytes(), lastId.toString().getBytes());
        ctx.putObject(ticketId.getBytes(), lastId.toString().getBytes());
        ctx.putObject(tickets.getBytes(), lastId.toString().getBytes());
        return Response.ok(lastId.toString().getBytes());
    }

    public Response startLuckDraw(Context ctx) {

        String caller = ctx.caller();
        if (caller.isEmpty()) {
            return Response.error("missing initiator");
        }
        String admin = new String(ctx.getObject(admin_key.getBytes()));

        if (caller != admin) {
            return Response.error("only the admin can add new asset type");
        }

        BigDecimal seed = new BigDecimal(new String(ctx.args().get("seed")));
        if (seed.compareTo(BigDecimal.ZERO) == 0) {
            return Response.error("missing seed");
        }

        BigDecimal lastId = new BigDecimal(new String(ctx.getObject(tickets.getBytes())));
        if (lastId.compareTo(BigDecimal.ZERO) == 0) {
            return Response.error("no luck draw tickets");
        }
        Random rander = new Random(seed.longValue()); //TODO truncate

        long luckid = (rander.nextLong() % lastId.longValue()) + 1;
        byte[] luckUser = ctx.getObject((ticketId + String.format("%d", luckid)).getBytes()); // TODO
        ctx.putObject(result.getBytes(), luckUser);
        return Response.ok(luckUser);
    }


    public Response getResult(Context ctx) {
        String luckUser = new String(ctx.getObject(result.getBytes()));
        if (luckUser.isEmpty()) {
            return Response.error("get luck draw result failed");
        }
        return Response.ok(luckUser.getBytes());
    }

    public static void main(String[] args) {
        Driver.serve(new LuckDraw());
    }
}